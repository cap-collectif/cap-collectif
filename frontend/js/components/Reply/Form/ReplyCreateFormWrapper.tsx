import * as React from 'react'

import styled from 'styled-components'
import { FormattedMessage } from 'react-intl'
import { connect } from 'react-redux'
import { createFragmentContainer, graphql } from 'react-relay'
import { Alert } from 'react-bootstrap'
import type { ReplyCreateFormWrapper_questionnaire } from '~relay/ReplyCreateFormWrapper_questionnaire.graphql'
import '~relay/ReplyCreateFormWrapper_questionnaire.graphql'
import ReplyForm from './ReplyForm'
import type { User } from '~/redux/modules/user'
import '~/redux/modules/user'
import type { GlobalState } from '~/types'
import { QuestionnaireStepPageContext } from '~/components/Page/QuestionnaireStepPage.context'
type Props = {
  readonly questionnaire: ReplyCreateFormWrapper_questionnaire
  readonly user: User | null | undefined
  readonly setIsShow: (show: boolean) => void
}
const ReplyCreateFormWrapperContainer = styled.div`
  margin-top: 35px;
`
export const ReplyCreateFormWrapper = ({ questionnaire, setIsShow }: Props) => {
  const { anonymousRepliesIds } = React.useContext(QuestionnaireStepPageContext)
  const formIsDisabled =
    questionnaire.contribuable &&
    questionnaire.viewerReplies &&
    questionnaire.viewerReplies.totalCount > 0 &&
    !questionnaire.multipleRepliesAllowed
  return (
    <ReplyCreateFormWrapperContainer>
      {
        formIsDisabled && (
          <Alert bsStyle="warning" className="hidden-print">
            <strong>
              <FormattedMessage id="reply.user_has_reply.reason" />
            </strong>
            <p>
              <FormattedMessage id="reply.user_has_reply.error" />
            </p>
          </Alert>
        )
      }
      <ReplyForm
        questionnaire={questionnaire}
        reply={null}
        setIsShow={setIsShow}
        anonymousRepliesIds={anonymousRepliesIds}
      />
    </ReplyCreateFormWrapperContainer>
  )
}

const mapStateToProps = (state: GlobalState) => ({
  user: state.user.user,
})

// @ts-ignore
const container = connect(mapStateToProps)(ReplyCreateFormWrapper)
export default createFragmentContainer(container, {
  questionnaire: graphql`
    fragment ReplyCreateFormWrapper_questionnaire on Questionnaire
    @argumentDefinitions(isAuthenticated: { type: "Boolean!" }, participantToken: { type: "String" }) {
      id
      multipleRepliesAllowed
      phoneConfirmationRequired
      contribuable
      viewerReplies @include(if: $isAuthenticated) {
        totalCount
      }
      ...ReplyForm_questionnaire @arguments(isAuthenticated: $isAuthenticated, participantToken: $participantToken)
    }
  `,
})
