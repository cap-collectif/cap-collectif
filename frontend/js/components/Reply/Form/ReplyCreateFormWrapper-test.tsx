/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import { $refType, $fragmentRefs } from '~/mocks'
import { ReplyCreateFormWrapper } from './ReplyCreateFormWrapper'

const baseProps = {
  user: {
    id: '23',
    username: 'Tom',
    isEmailConfirmed: true,
    isPhoneConfirmed: true,
    phone: '0606060606',
    isAdmin: false,
    isSuperAdmin: false,
    isProjectAdmin: false,
    isOrganizationMember: false,
    isOnlyProjectAdmin: false,
    isEvaluerOnLegacyTool: false,
    isEvaluerOnNewTool: false,
    email: 'test@gmail.com',
    newEmailToConfirm: null,
    media: null,
    roles: [],
    displayName: 'TomTom',
    uniqueId: '234',
    _links: {
      profile: 'http://test.com',
    },
  },
  questionnaire: {
    ' $refType': $refType,
    ' $fragmentRefs': $fragmentRefs,
    id: '123',
    contribuable: true,
    multipleRepliesAllowed: true,
    viewerReplies: {
      totalCount: 1,
    },
    phoneConfirmationRequired: false,
  },
  setIsShow: jest.fn(),
}
const props = {
  basic: baseProps,
  withoutMultipleVotes: {
    ...baseProps,
    questionnaire: {
      ...baseProps.questionnaire,
      multipleRepliesAllowed: false,
      viewerReplies: {
        totalCount: 2,
      },
    },
  },
  notContribuable: {
    ...baseProps,
    questionnaire: { ...baseProps.questionnaire, contribuable: false, multipleRepliesAllowed: false },
  },
  notLoginUser: { ...baseProps, user: null },
  anonymousParticipantsAllowed: {
    ...baseProps,
    questionnaire: {
      ...baseProps.questionnaire,
    },
  },
}
describe('<ReplyCreateFormWrapper />', () => {
  it('should render no alert an an enabled form', () => {
    const wrapper = shallow(<ReplyCreateFormWrapper {...props.basic} />)
    expect(wrapper).toMatchSnapshot()
  })
  it('should render no alert when anonymous particpants are allowed', () => {
    const wrapper = shallow(<ReplyCreateFormWrapper {...props.anonymousParticipantsAllowed} />)
    expect(wrapper).toMatchSnapshot()
  })
  it('should render an alert an a disabled form when form is contribuable and user is not logged in', () => {
    const wrapper = shallow(<ReplyCreateFormWrapper {...props.notLoginUser} />)
    expect(wrapper).toMatchSnapshot()
  })
  it("should render an alert an a disabled form when form is contribuable and doesn't allow multiple votes and user has already votes", () => {
    const wrapper = shallow(<ReplyCreateFormWrapper {...props.withoutMultipleVotes} />)
    expect(wrapper).toMatchSnapshot()
  })
  it('should no alert an a disabled form when form is not contribuable', () => {
    const wrapper = shallow(<ReplyCreateFormWrapper {...props.notContribuable} />)
    expect(wrapper).toMatchSnapshot()
  })
})
