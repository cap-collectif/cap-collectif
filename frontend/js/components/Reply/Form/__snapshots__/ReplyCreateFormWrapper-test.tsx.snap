// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReplyCreateFormWrapper /> should no alert an a disabled form when form is not contribuable 1`] = `
<ReplyCreateFormWrapper__ReplyCreateFormWrapperContainer>
  <Relay(withRouter(Connect(injectIntl(ReduxForm))))
    anonymousRepliesIds={Array []}
    questionnaire={
      Object {
        " $fragmentRefs": null,
        " $refType": null,
        "contribuable": false,
        "id": "123",
        "multipleRepliesAllowed": false,
        "phoneConfirmationRequired": false,
        "viewerReplies": Object {
          "totalCount": 1,
        },
      }
    }
    reply={null}
    setIsShow={[MockFunction]}
  />
</ReplyCreateFormWrapper__ReplyCreateFormWrapperContainer>
`;

exports[`<ReplyCreateFormWrapper /> should render an alert an a disabled form when form is contribuable and doesn't allow multiple votes and user has already votes 1`] = `
<ReplyCreateFormWrapper__ReplyCreateFormWrapperContainer>
  <Alert
    bsClass="alert"
    bsStyle="warning"
    className="hidden-print"
    closeLabel="Close alert"
  >
    <strong>
      <FormattedMessage
        id="reply.user_has_reply.reason"
        values={Object {}}
      />
    </strong>
    <p>
      <FormattedMessage
        id="reply.user_has_reply.error"
        values={Object {}}
      />
    </p>
  </Alert>
  <Relay(withRouter(Connect(injectIntl(ReduxForm))))
    anonymousRepliesIds={Array []}
    questionnaire={
      Object {
        " $fragmentRefs": null,
        " $refType": null,
        "contribuable": true,
        "id": "123",
        "multipleRepliesAllowed": false,
        "phoneConfirmationRequired": false,
        "viewerReplies": Object {
          "totalCount": 2,
        },
      }
    }
    reply={null}
    setIsShow={[MockFunction]}
  />
</ReplyCreateFormWrapper__ReplyCreateFormWrapperContainer>
`;

exports[`<ReplyCreateFormWrapper /> should render an alert an a disabled form when form is contribuable and user is not logged in 1`] = `
<ReplyCreateFormWrapper__ReplyCreateFormWrapperContainer>
  <Relay(withRouter(Connect(injectIntl(ReduxForm))))
    anonymousRepliesIds={Array []}
    questionnaire={
      Object {
        " $fragmentRefs": null,
        " $refType": null,
        "contribuable": true,
        "id": "123",
        "multipleRepliesAllowed": true,
        "phoneConfirmationRequired": false,
        "viewerReplies": Object {
          "totalCount": 1,
        },
      }
    }
    reply={null}
    setIsShow={[MockFunction]}
  />
</ReplyCreateFormWrapper__ReplyCreateFormWrapperContainer>
`;

exports[`<ReplyCreateFormWrapper /> should render no alert an an enabled form 1`] = `
<ReplyCreateFormWrapper__ReplyCreateFormWrapperContainer>
  <Relay(withRouter(Connect(injectIntl(ReduxForm))))
    anonymousRepliesIds={Array []}
    questionnaire={
      Object {
        " $fragmentRefs": null,
        " $refType": null,
        "contribuable": true,
        "id": "123",
        "multipleRepliesAllowed": true,
        "phoneConfirmationRequired": false,
        "viewerReplies": Object {
          "totalCount": 1,
        },
      }
    }
    reply={null}
    setIsShow={[MockFunction]}
  />
</ReplyCreateFormWrapper__ReplyCreateFormWrapperContainer>
`;

exports[`<ReplyCreateFormWrapper /> should render no alert when anonymous particpants are allowed 1`] = `
<ReplyCreateFormWrapper__ReplyCreateFormWrapperContainer>
  <Relay(withRouter(Connect(injectIntl(ReduxForm))))
    anonymousRepliesIds={Array []}
    questionnaire={
      Object {
        " $fragmentRefs": null,
        " $refType": null,
        "contribuable": true,
        "id": "123",
        "multipleRepliesAllowed": true,
        "phoneConfirmationRequired": false,
        "viewerReplies": Object {
          "totalCount": 1,
        },
      }
    }
    reply={null}
    setIsShow={[MockFunction]}
  />
</ReplyCreateFormWrapper__ReplyCreateFormWrapperContainer>
`;
